File: neet_code_150_cpp/heap_p_queue/last_stone_weight.cpp
#include <iostream>
#include <vector>
#include <queue>

// AC
class Solution {
    public:
        int lastStoneWeight(std::vector<int>& stones) {
            std::priority_queue<int> max_heap(stones.begin(), stones.end());

            while(max_heap.size() > 1) {
                auto y = max_heap.top();
                max_heap.pop();
                auto x = max_heap.top();
                max_heap.pop();

                if(x != y) {
                    max_heap.push(y - x);
                }
            }

            if(max_heap.size() == 0) {
                return 0;
            }

            return max_heap.top();
        }
};

// 模範解答
class SolutionAns {
    public:
        int lastStoneWeight(std::vector<int>& stones) {
            std::priority_queue<int> pq;
            for(auto& v: stones) {
                pq.push(v);
            }

            while(pq.size() > 1) {
                int y = pq.top();
                pq.pop();
                int x = pq.top();
                pq.pop();

                if(y > x) {
                    pq.push(y - x);
                }
            }

            if(pq.empty()) {
                return 0;
            }

            return pq.top();
        }
};

int main(void) {
    auto case_1 = std::vector{2, 7, 4, 1, 8, 1};
    auto case_2 = std::vector{1};

    Solution s_1;
    std::cout << s_1.lastStoneWeight(case_1) << std::endl;
    std::cout << s_1.lastStoneWeight(case_2) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.lastStoneWeight(case_1) << std::endl;
    std::cout << s_ans.lastStoneWeight(case_2) << std::endl;
}

File: neet_code_150/src/heap_p_queue/last_stone_weight.rs
use std::collections::BinaryHeap;

// AC
struct Solution {}
impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut max_heap: BinaryHeap<i32> = BinaryHeap::from(stones);

        while max_heap.len() > 1 {
            let y = max_heap.peek().unwrap().clone();
            max_heap.pop();
            let x = max_heap.peek().unwrap().clone();
            max_heap.pop();

            if x != y {
                max_heap.push(y - x);
            }
        }

        if max_heap.is_empty() {
            return 0;
        }
        *max_heap.peek().unwrap()
    }
}

// 模範解答
struct SolutoinAns {}
impl SolutoinAns {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut sotone_heap = BinaryHeap::new();
        for stone in stones {
            sotone_heap.push(stone);
        }

        while sotone_heap.len() > 1 {
            let first = sotone_heap.pop().unwrap();
            let second = sotone_heap.pop().unwrap();
            sotone_heap.push(first - second);
        }

        match sotone_heap.peek() {
            Some(val) => *val,
            None => 0,
        }
    }
}

fn main() {
    let case_1 = vec![2, 7, 4, 1, 8, 1];
    // => 1
    let case_2 = vec![1];
    // => 1

    println!("case_1: {}", Solution::last_stone_weight(case_1.clone()));
    println!("case_2: {}", Solution::last_stone_weight(case_2.clone()));
}
